### ISSUE TITLE
Singleton DefaultMethodInfoBasedNodeTypeRegistryFactory is not thread-safe

### ISSUE THREAD

**Issue by KevinCathcart** [aspnet/EntityFrameworkCore#12682]
`DefaultMethodInfoBasedNodeTypeRegistryFactory` is a registered as singleton instance (However we are really only concerned about its base class). The base class `MethodInfoBasedNodeTypeRegistryFactory`, accepts a ReLinq `MethodInfoBasedNodeTypeRegistry` instance which from a quick glance at the source is pretty clearly not thread safe is mutating members are used.

The factory singeton gets injected into every scoped lifetime `QueryModelGenerator` which immediately calls `Create()` on it. `MethodInfoBasedNodeTypeRegistry.Create()` in turn calls `.Register()` on the stored MethodInfoBasedNodeTypeRegistry field which is a mutating member.

It looks like a `QueryModelGenerator` is an indirect constructor injected dependency of `DbContext` via `QueryModelGenetator` -> `QueryCompiler` -> `EntityQueryProvider`-> `DbContextDependencies` -> `DbContext`.

So it looks to me like every time a new `DbContext` is created in a fresh scope (plausibly in a different thread), we end up mutating an effectively static object that is not thread safe if mutated! 

It is possible that I overlooking something that would make this safe, but I don't immediately see it. 

I came across this when investigating an `System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.` error that appeared to stem from a `MethodInfoBasedNodeTypeRegistry` being used by EFCore.

**Comment by benaadams** [comment]
Added an issue to corefx so its easier to find in search https://github.com/dotnet/corefx/issues/31186

**Comment by fschlaef** [comment]
(This comment was originally in #12713 but it belongs here)

I have the exact same exception occurring at random despite not using Dictionary at all.

```
System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Remotion.Linq.Parsing.Structure.NodeTypeProviders.MethodInfoBasedNodeTypeRegistry.GetNodeType(MethodInfo method)
   at Remotion.Linq.Parsing.Structure.NodeTypeProviders.MethodInfoBasedNodeTypeRegistry.IsRegistered(MethodInfo method)
   at Remotion.Linq.Parsing.Structure.NodeTypeProviders.CompoundNodeTypeProvider.<>c__DisplayClass4_0.<IsRegistered>b__0(INodeTypeProvider p)
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Remotion.Linq.Parsing.Structure.NodeTypeProviders.CompoundNodeTypeProvider.IsRegistered(MethodInfo method)
   at Remotion.Linq.Parsing.Structure.ExpressionTreeParser.GetQueryOperatorExpression(Expression expression)
   at Remotion.Linq.Parsing.ExpressionVisitors.SubQueryFindingExpressionVisitor.Visit(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Remotion.Linq.Parsing.ExpressionVisitors.SubQueryFindingExpressionVisitor.Visit(Expression expression)
   at Remotion.Linq.Parsing.ExpressionVisitors.SubQueryFindingExpressionVisitor.Process(Expression expressionTree, INodeTypeProvider nodeTypeProvider)
   at Remotion.Linq.Parsing.Structure.MethodCallExpressionParser.ProcessArgumentExpression(Expression argumentExpression)
   at System.Linq.Enumerable.SelectListPartitionIterator`2.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Remotion.Linq.Parsing.Structure.MethodCallExpressionParser.Parse(String associatedIdentifier, IExpressionNode source, IEnumerable`1 arguments, MethodCallExpression expressionToParse)
   at Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseMethodCallExpression(MethodCallExpression methodCallExpression, String associatedIdentifier)
   at Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseNode(Expression expression, String associatedIdentifier)
   at Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseMethodCallExpression(MethodCallExpression methodCallExpression, String associatedIdentifier)
   at Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseTree(Expression expressionTree)
   at Remotion.Linq.Parsing.Structure.QueryParser.GetParsedQuery(Expression expressionTreeRoot)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryModelGenerator.ParseQuery(Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass13_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at MyProject.Controllers.Parameters.CurrencyController.GetList() in C:\projects\MyProject\src\MyProject\Controllers\Parameters\CurrencyController.cs:line 223
   at MyProject.Controllers.Parameters.ParameterController.GetParameters() in C:\projects\MyProject\src\MyProject\Controllers\Parameters\ParameterController.cs:line 168
   at lambda_method(Closure , Object , Object[] )
   at Microsoft.Extensions.Internal.ObjectMethodExecutor.Execute(Object target, Object[] parameters)
   at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.SyncActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at MyProject.Startup.<>c.<<Configure>b__6_1>d.MoveNext() in C:\projects\MyProject\src\MyProject\Startup.cs:line 179
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```
Relevant functions :

CurrencyController.GetList() :
```csharp
public List<dynamic> GetList()
        {
            return _context.Currency
                .Where(a => a.Enabled)
                .Select(c => new
                {
                    c.Id,
                    c.Designation,
                    c.Enabled,
                    c.Code,
                    Name = $"{c.Code} - {c.Designation}"
                }).ToList<dynamic>();
        }
```

ParameterController.GetParameters() :
```csharp
[HttpGet]
public IActionResult GetParameters()
{
	List<dynamic> result = new List<dynamic>();

	// Currency
	string currency = _currencyController.GetType();
	if(moduleAccessLevels[currency])
	{
		result.Add(GetParameterObject(currency, _currencyController.GetList()));
	}
	
	// We do this with a bunch of other controllers
	
	return _responseService.Success(result);
}

private dynamic GetParameterObject(string moduleName, List<dynamic> data)
{
	return new
	{
		ParameterName = moduleName,
		List = data
	};
}
```

Additional information :

- The exception is completely random. Sometimes we don't see it for days, then it throws three times in a row
- It _seems_ that it only throws on Windows 10. I have never seen it on a Windows 7 machine, but take this with a huge pinch of salt as this exception is random
- I cannot be 100% sure, but I think I've seen the same exception since 2.1.0
- Sometimes it throws on really simple code, like ```_context.Something.Where().Select().ToList()``` which is strange
- GetParameters() used to call static functions inside controllers and passing them a DbContext. I though this was the culprit so I switched to injected controllers, but it didn't do the trick.

Hope you can make something out of this.

**Comment by divega** [comment]
**EF Triage:** Once we understand the fix we will consider if this needs to be patched.

**Comment by smitpatel** [comment]
@MichaelKetting - Question for Re-Linq query parser end. We use `INodeTypeProvider` in following way.
```C#
private QueryParser CreateQueryParser(INodeTypeProvider nodeTypeProvider)
            => new QueryParser(
                new ExpressionTreeParser(
                    nodeTypeProvider,
                    new CompoundExpressionTreeProcessor(
                        new IExpressionTreeProcessor[]
                        {
                            new PartialEvaluatingExpressionTreeProcessor(_evaluatableExpressionFilter),
                            new TransformingExpressionTreeProcessor(ExpressionTransformerRegistry.CreateDefault())
                        })));
```
So the question would be, should each QueryParser needs a new `nodeTypeProvider` or we can share the same instance across different QueryParsers?

**Comment by MichaelKetting** [comment]
@smitpatel I'm using a single instance of QueryParser, too. And yes, we should have made the threading-constraints more explicit. Sorry about that.

Looking at things, yes, while the interface does not offer mutating features, the implementation type is a different beast. We probably should probably update this to a builder + immutable data structure pattern to make threadsafety more explicit. (https://www.re-motion.org/jira/browse/RMLNQ-123)

As for the specific error: re-linq itself uses MethodInfoBasedNodeTypeRegistry.CreateFromRelinqAssembly() which is the only spot where the MethodInfoBasedNodeTypeRegistry instance is first instantiated and then mutated before the instance is shared with the rest of the code.

If EntityFramework is exposing the MethodInfoBasedNodeTypeRegistry.Register(...) method after the single-threaded composition, than yes, that would most definitely be an issue. I did a very quick peek into the EF sources, and yes, if MethodInfoBasedNodeTypeRegistryFactory.Create() is called multiple times, than that will be an issue since it would write to the dictionary while the dictionary is used on other threads. There is also MethodInfoBasedNodeTypeRegistryFactory.RegisterMethods, which has the same problem.

So, to sum it up, the implementation of the MethodInfoBasedNodeTypeRegistryFactory probably needs to be updated so the MethodInfoBasedNodeTypeRegistry will only be muted during initialization. Once you hold a reference to the field, it must be of the INodeTypeProvider interface to ensure that no one's using the Register() methods. Creating a new instance of the MethodInfoBasedNodeTypeRegistry would not be recommended for performance reasons.

Best regards, Michael

**Comment by ajcvickers** [comment]
#### Description 

A non-thread safe dictionary in Relinq is accessed by multiple threads concurrently during service initialization in relational providers.

#### Customer Impact 

2.1: Occasional exceptions due to this corefx change: https://github.com/dotnet/corefx/issues/28123
Pre 2.0: Presumably even rarer failures due to the unsafe access, but nothing ever reported.

#### Regression? 

 Technically, no. Effectively, yes, since 2.1 apps are now more likely to fail than 2.0 apps.

#### Risk 

Low. The change isn't as simple as using a ConcurrentDictionary due to the way it is used in D.I. service resolution and because the dictionary is located in Relinq code. So, the fix is to re-plum the initialization here, but in a way that doesn't break any existing providers.

**Comment by vivmishra** [comment]
Approved for 2.1.5

**Comment by fschlaef** [comment]
Question: why this moved to 2.1.5 while the other "non-concurrent" bug remained in 2.1.4 ?

**Comment by ajcvickers** [comment]
@fschlaef 2.1.4 is closing down for new work; this issue was not fixed in time to ship in 2.1.4.

**Comment by smitpatel** [comment]
Link to PR https://github.com/aspnet/EntityFrameworkCore/pull/12778

**Comment by ajcvickers** [comment]
@smitpatel The branch is now ready for this to be merged.
